src/transformers/generation/utils.py

```diff
@@ -2792,24 +2775,12 @@ def _sample(
             next_token_scores = logits_processor(input_ids, next_token_logits)
 
             # Store scores, attentions and hidden_states when required
-            if return_dict_in_generate:
-                if output_scores:
-                    scores += (next_token_scores,)
-                if output_logits:
-                    raw_logits += (next_token_logits,)
-                if output_attentions:
-                    decoder_attentions += (
-                        (outputs.decoder_attentions,) if self.config.is_encoder_decoder else (outputs.attentions,)
-                    )
-                    if self.config.is_encoder_decoder:
-                        cross_attentions += (outputs.cross_attentions,)
-
-                if output_hidden_states:
-                    decoder_hidden_states += (
-                        (outputs.decoder_hidden_states,)
-                        if self.config.is_encoder_decoder
-                        else (outputs.hidden_states,)
-                    )
+            self._accumulate_generate_output(
+                generate_output,
+                score_tensors=next_token_scores if generate_output["output_scores"] else None,
+                logits_tensors=next_token_logits if generate_output["output_logits"] else None,
```

Comment:
suggestion: always pass `next_token_scores` and `next_token_logits`, handle whether they are used or not inside the function

(less logic for the decoding method)